src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+//! they claim it has.
+//! The submission is a vector of field elements laid out as follows:
+//! ```text
+//! |---- bits_per_entry * entries ----|---- bits_for_norm ----|
+//!  ^                                  ^
+//!  \- the input vector entries        |
+//!                                     \- the encoded norm
+//! ```
+//!
+//! ### Validity
+//!
+//! In addition to checking that every submission entry is `0` or `1`, the validation
+//! circuit of this type computes the norm and compares to what the client
+//! claimed.
+//!
+//! ### Value `1`
+//!
+//! We actually do not allow the submitted norm or vector entries to be
+//! exactly `1`, but rather require them to be strictly less. Supporting `1` would
+//! entail a more fiddly encoding and is not necessary for our usecase.
+
+pub mod compatible_float;
+
+use crate::field::{FieldElement, FieldElementExt};
+use crate::flp::gadgets::{BlindPolyEval, ParallelSumGadget, PolyEval};
+use crate::flp::types::fixedpoint_l2::compatible_float::CompatibleFloat;
+use crate::flp::{FlpError, Gadget, Type};
+use crate::polynomial::poly_range_check;
+use fixed::traits::Fixed;
+
+use std::{convert::TryFrom, convert::TryInto, fmt::Debug, marker::PhantomData};
+
+/// The fixed point vector sum data type. Each measurement is a vector of fixed point numbers of
+/// type `T`, and the aggregate result is the float vector of the sum of the measurements.
+///
+/// The validity circuit verifies that the L2 norm of each measurement is bounded by 1.
+///
+/// The [*fixed* crate](https://crates.io/crates/fixed) is used for fixed point numbers, in
+/// particular, exactly the following types are supported:
+/// `FixedI16<U15>`, `FixedI32<U31>` and `FixedI64<U63>`.
+///
+/// Depending on the size of the vector that needs to be transmitted, a corresponding field type has
+/// to be chosen for `F`. For a `n`-bit fixed point type and a `d`-dimensional vector, the field
+/// modulus needs to be larger than `d * 2^(2n-2)` so there are no overflows during norm validity
+/// computation.
+#[derive(Debug, PartialEq, Eq)]
+pub struct FixedPointBoundedL2VecSum<
+    T: Fixed,
+    F: FieldElement,
+    SPoly: ParallelSumGadget<F, PolyEval<F>>,
+    SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>>,
+> {
+    bits_per_entry: usize,
+    fi_bits_per_entry: F::Integer,
+    entries: usize,
+    bits_for_norm: usize,
+    range_01_checker: Vec<F>,
+    norm_summand_poly: Vec<F>,
+    phantom: PhantomData<(T, SPoly, SBlindPoly)>,
+    // range/position constants
+    range_norm_begin: usize,
+    range_norm_end: usize,
+
+    // configuration of parallel sum gadgets
+    gadget0_calls: usize,
+    gadget0_chunk_len: usize,
+    gadget1_calls: usize,
+    gadget1_chunk_len: usize,
+}
+
+impl<
+        T: Fixed,
+        F: FieldElement,
+        SPoly: ParallelSumGadget<F, PolyEval<F>>,
+        SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>>,
+    > FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>
```

Comment:
Readability: Move trait bounds to the `where` block below
```suggestion
impl<T, F, SPoly, SBlindPoly>
```